Wood

Petermarcu opened the pull request:
Ignoring the MSBuild conversion noise, I'd be interested in feedback on my solution for the different blocks that represent wood. Kind of an ugly situation but I opted for the developer only interacting with "Wood" and hiding that that is actually possibly two different block types in the de/serialization.

Petermarcu committed:
converting to msbuild

Petermarcu committed:
Merge remote-tracking branch 'upstream/master'

Petermarcu committed:
add support for other woods

bleroy commented on Decent.Minecraft.Client/Java/JavaBlock.cs:
```diff
@@ -80,7 +80,9 @@ static JavaBlock()
                 (StoneBricks)new ChiseledStoneBricks();
-            _ctors[Id<Wood>()] = d => new Wood((WoodSpecies)(d & 0x3), (Axis)(d & 0xC));
+            // Wood is both 17 and 162 depending on species
+            _ctors[17] = d => new Wood((WoodSpecies)(d & 0x3), (Axis)(d & 0xC));
```
Look at the precedent of `Snow` and `SnowLayer`: the main one, `Snow`, is the only one that has an actual type on our side, and it's in the Java registry. The other one, `SnowLayer`, is defined as a constant so we don't have a magic number hanging in the serialization code. I think this is pretty much analogous to what you're trying to do with wood.

Petermarcu replied:
Yeah, happy to move the magic number out into a constant. Sounds like you are in agreement that having a single wood type to represent all woods makes sense. The challenge with wood is that it has more than 4 types and overflowed into another block type. AcaciaWood isn't accurate because its also Dark Oak and will probably have another one added at some point in the future. Seems like a general pattern for things that have more than 4 types to be represented by different block ID's and the different ranges in the Enum values need to map to different block ID's.

Are you sure Snow is implemented correctly? Snow and Snow layer are different block types and have different characteristics. An 8 layer SnowLayer block isn't the same as a solid Snow Block. One Melts and the other doesn't.

bleroy replied:
Oh yes, I'm very much in agreement. That's what I wanted all along ;) That sort of thing is all over: it really shows how MC was grown organically. Our class hierarchy should be more rational than that.

On snow, the behavior is one thing, thickness is another. I'll check later, but from what I remember, the wiki spec and reality were out of sync, and the value 8 wasn't actually possible except through a snow block.

bleroy reviewed (changes requested):
I'd leave `Id<Wood>()` in, and introduce a constant for `AcaciaWood`.

Petermarcu committed:
fixup wood and add identify

Petermarcu commented:
Ok, I updated to do the constants and added a mode in the app that prints out what type and id a block was serialized into. I'm a little concerned about it getting confusing that we're mapping multiple block types to one. Thinking we should add some type of ToString method to each block type that will print out its properties so we can see the details when we hit a block :)

Petermarcu committed:
fixing constants

Petermarcu committed:
fixing up comment and whitespace

bleroy commented:
Agreed about ToString.
